### PR TITLE
Make Input accept json

### LINKED ISSUE #635: Make Input accept JSON payload
It would easier for users if they can shove case classes into an `Input` payload as JSON objects. I'm having something like this in mind (pseudocode):

```scala
def withJsonBody[A](a: A)(implicit e: Encode.Json[A], cs: Option[Charset] = Som("utf8")): Input =
  withBody(Buf.Text(e(a, cs)), Some("application/json;charset=" + cs))
```

### PR DESCRIPTION
**Pull request by BenFradet**
fixes #635

### CODE REVIEW

File: core/src/main/scala/io/finch/Input.scala [hunk 1]
```diff
@@ -32,6 +34,19 @@ final case class Input(request: Request, path: Seq[String]) {
   }
 
   /**
+   * Overrides (mutates) the payload of this input in order to reflect `a` as json and returns
+   * `this`.
+   *
+   * @note application/json will be passed as content type as well as the charset used to encode
+   *       the json
+   */
+  def withJson[A](a: A)(implicit e: Encode[A], cs: Option[Charset] = Some(Charsets.Utf8)): Input =
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: It should be `Encode.Json[A]` so we provide compile-time guarantee on using JSON encoders here.

File: core/src/main/scala/io/finch/Input.scala [hunk 2]
```diff
@@ -32,6 +34,19 @@ final case class Input(request: Request, path: Seq[String]) {
...
+    cs match {
+      case Some(charset) => withBody(e(a, charset), Some(s"application/json;charset=$charset"))
+      case None => withBody(e(a, Charsets.Utf8), Some(s"application/json;charset=utf-8"))
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: Let's not set `charset=` when it's `None`.

File: docs/endpoint.md [hunk 3]
```diff
@@ -632,19 +632,20 @@ val bar: Input = Input.post("/bar")
...
-import io.circe.syntax._
+import io.circe.generic.auto._
 import io.finch._
-import com.twitter.io.Buf
+import io.finch.circe._
 
-val baz: Input = Input.put("/baz")
-  .withBody(Buf.Utf8(Map("a" -> "b").asJson.toString), Some("application/json;charset=utf8"))
+case class Baz(m: Map[String, String])
+val baz: Input = Input.put("/baz").withJson(Baz(Map("a" -> "b")))
```
Comment (vkostyukov) [inline docs/endpoint.md]: This is pretty sweet!

Comment (BenFradet) [inline docs/endpoint.md]: yup!

File: core/src/main/scala/io/finch/Input.scala [hunk 4]
```diff
@@ -32,6 +34,20 @@ final case class Input(request: Request, path: Seq[String]) {
...
 
   /**
+   * Overrides (mutates) the payload of this input in order to reflect `a` as json and returns
+   * `this`.
+   *
+   * @note application/json will be passed as content type as well as the charset used to encode
+   *       the json
+   */
+  def withJson[A](a: A)(
+      implicit e: Encode.Json[A], cs: Option[Charset] = Some(Charsets.Utf8)): Input =
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/Input.scala]: `cs: Option[Charset]` should not be implicit. You want to get something like:

```scala
def withJson[A](a: A, cs: Option[Charset] = Some(Charsets.Utf8))(implicit e: Encode.Json[A]): Input =
```

File: core/src/main/scala/io/finch/Input.scala [hunk 5]
```diff
@@ -32,6 +34,20 @@ final case class Input(request: Request, path: Seq[String]) {
...
+    cs match {
+      case Some(charset) => withBody(e(a, charset), Some(s"application/json;charset=$charset"))
+      case None => withBody(e(a, Charsets.Utf8), Some(s"application/json"))
```
Comment (yanana) [inline core/src/main/scala/io/finch/Input.scala]: You don't need string interpolation here.

**Comment by BenFradet** [comment]
updated :+1: thanks for the review @vkostyukov @yanana

File: core/src/main/scala/io/finch/Input.scala [hunk 6]
```diff
@@ -32,6 +34,20 @@ final case class Input(request: Request, path: Seq[String]) {
...
+   * Overrides (mutates) the payload of this input in order to reflect `a` as json and returns
+   * `this`.
+   *
+   * @note application/json will be passed as content type as well as the charset used to encode
+   *       the json
+   */
+  def withJson[A](a: A, charset: Option[Charset] = Some(Charsets.Utf8))(
+      implicit e: Encode.Json[A]): Input =
+    charset match {
+      case Some(cs) => withBody(e(a, cs), Some(s"application/json;charset=$cs"))
```
Comment (yanana) [inline core/src/main/scala/io/finch/Input.scala]: Should we put a space after the semicolon?

Comment (BenFradet) [inline core/src/main/scala/io/finch/Input.scala]: it's not done anywhere else in the code base so I don't think so, no.

Comment (yanana) [inline core/src/main/scala/io/finch/Input.scala]: I see.

**Comment by vkostyukov** [comment]
Not sure why Codecov complains (-0.1% coverage?), but going to go ahead and merge it. Thanks @BenFradet!

**Pull request merged**